### PR TITLE
Add `gulp-imagemin` task

### PR DESCRIPTION
**Pull request by claire-codes**
When the task `gulp imagemin` is run, it finds all the images in the `web/images` directory within each theme in `app/design` and minifies them if possible.

* The files are minified in place
* There is a default config provided which can be overwritten by create a `imageMinConfig.json` file in the normal place in `dev/tools`.
* Package used is [https://github.com/sindresorhus/gulp-imagemin](https://github.com/sindresorhus/gulp-imagemin)

I would have liked to add a caching task that would only attempt to minify images that are new or updated but I couldn't work out how to get it to work with frontools helper, tasks and plugins ecosystem.

I would also like to be able to print out a message after the images have finished minifying per theme.

### CODE REVIEW

File: helper/imagemin.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+          imageDir       = srcBase + '/web/images/**/*',
+          minnedImageDir = srcBase + '/web/images/';
+
+    let imageMinConfig;
+
+    try {
+        // Checks for project specific config in dev/tools/frontools/config
+        imageMinConfig = require('../helper/config-loader')('imageMinConfig.json', plugins, config);
+    } catch (e) {
+        imageMinConfig = { "optimizationLevel": 3, "progressive": true, "interlaced": true };
```
Comment (ghost) [inline helper/imagemin.js]: We don't need to do this try/catch, if you check in the `config-loader` you'll see it throws an error if the config isn't found.

We should create a .sample config in the `config/` directory with the recommended defaults, it'll get copied over to the project on `gulp setup` and if it's then manually removed the `config-loader` module will throw out the error.

File: helper/imagemin.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+          // Look for any images in the web/images directory and minify them in place
+          imageDir       = srcBase + '/web/images/**/*',
+          minnedImageDir = srcBase + '/web/images/';
+
+    let imageMinConfig;
+
+    try {
+        // Checks for project specific config in dev/tools/frontools/config
+        imageMinConfig = require('../helper/config-loader')('imageMinConfig.json', plugins, config);
```
Comment (ghost) [inline helper/imagemin.js]: We can go with `imagemin.json` to keep to the current naming convention used by the other config files.

File: helper/imagemin.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
+'use strict';
+module.exports = function(gulp, plugins, config, name) { // eslint-disable-line func-names
+  // Return function that is executed inside of .pipe()
+  return () => {
+    const theme          = config.themes[name],
+          srcBase        = config.projectPath + theme.src,
+
+          // Look for any images in the web/images directory and minify them in place
+          imageDir       = srcBase + '/web/images/**/*',
+          minnedImageDir = srcBase + '/web/images/';
```
Comment (ghost) [inline helper/imagemin.js]: Do you think it's worth running this on `/web/icons` too? I'm not sure how/if that'll impact the icon task... I seem to recall that gulpicon can only optimise PNGs.

File: helper/imagemin.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    try {
+        // Checks for project specific config in dev/tools/frontools/config
+        imageMinConfig = require('../helper/config-loader')('imageMinConfig.json', plugins, config);
+    } catch (e) {
+        imageMinConfig = { "optimizationLevel": 3, "progressive": true, "interlaced": true };
+    }
+
+    return gulp.src(imageDir)
+        .pipe(plugins.imagemin(imageMinConfig))
+        .pipe(gulp.dest(minnedImageDir));
```
Comment (ghost) [inline helper/imagemin.js]: If you want to log a message once the task is finished, you can pipe to the logger here (see L20 `helpers/css-lint.js` for an example).

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
>I would have liked to add a caching task

@claireparker: Have you got a plugin in mind that you'd like to use?

File: helper/imagemin.js (hunk 1 above)
Comment (claire-codes) [reply helper/imagemin.js]: Aaaah I didn't realise how that it is how it works 👍

**Comment by claire-codes** [comment]
@jakestubbz I wanted to use plain old `gulp-cache` as in this example https://markgoodyear.com/2014/01/getting-started-with-gulp/ (go to the "Compress Images" paragraph) which I've managed to set up on projects with less complex gulp files 😑 but couldn't get it to work properly with this setup.

File: helper/imagemin.js (hunk 3 above)
Comment (claire-codes) [reply helper/imagemin.js]: As to running it on different directories, I was thinking that we could run it on this directory by default and then optionally check for an argument to specify a different folder?

I've tested this on an SVG and it did manage to minify it from 4kb to 2kb! And the repo says "Minify PNG, JPEG, GIF and SVG images with imagemin"

**Review by ghost — APPROVED** [review]

**Comment by ghost** [comment]
@claireparker: Okay, cool. I'll have a read and see if I can figure out how we can plug it into frontools - it'd be useful to not have it running unless required.

**Comment by claire-codes** [comment]
@jakestubbz I've updated to allow a flag which provides a directory of the user's choice to minify. This means that you can do

```
gulp imagemin --dir dev/tests/backstop_data/bitmaps_reference/
```

to minify all the Backstop screenshots!

Still outstanding is adding a caching task here - currently it tries to minify everything each time the task is run.

**Pull request merged**